### PR TITLE
Dict: Use Ref Local Reassignment (C# 7.3)

### PR DESCRIPTION
**Pull request by benaadams**
```
Total bytes of diff: -634 (-0.02% of base)
    diff is an improvement.

Total byte diff includes 0 bytes from reconciling methods
        Base had    0 unique methods,        0 unique bytes
        Diff had    0 unique methods,        0 unique bytes

Top file improvements by size (bytes):
        -634 : System.Private.CoreLib.dasm (-0.02% of base)

1 total files with size differences (1 improved, 0 regressed), 0 unchanged.

Top method improvements by size (bytes):
         -96 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(ref,struct,ubyte):bool:this (4 methods)
         -90 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(ref,int,ubyte):bool:this (3 methods)
         -54 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(struct,struct,ubyte):bool:this (2 methods)
         -54 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(int,ref,ubyte):bool:this (3 methods)
         -36 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(struct,ref,ubyte):bool:this (3 methods)
         -36 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(long,ref,ubyte):bool:this (2 methods)
         -30 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(ref,ref,ubyte):bool:this
         -30 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(ref,bool,ubyte):bool:this
         -30 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(ref,long,ubyte):bool:this
         -30 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(ref,ushort,ubyte):bool:this
         -23 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(long,short,ubyte):bool:this
         -23 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(long,bool,ubyte):bool:this
         -23 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(int,ubyte,ubyte):bool:this
         -23 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(int,int,ubyte):bool:this
         -17 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(short,long,ubyte):bool:this
         -17 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(ushort,ref,ubyte):bool:this
         -11 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(struct,int,ubyte):bool:this
         -11 : System.Private.CoreLib.dasm - Dictionary`2:TryInsert(struct,ubyte,ubyte):bool:this

18 total methods with size differences (18 improved, 0 regressed), 17274 unchanged.
```

/cc @jkotas is moving to C# 7.3 a permitted change currently?

### CODE REVIEW

**Comment by benaadams** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline

**Comment by jkotas** [comment]
> moving to C# 7.3 a permitted change currently?

Once the tree opens for 2.2 (it should happen in a week).

File: src/mscorlib/System.Private.CoreLib.csproj [hunk 1]
```diff
@@ -39,7 +39,7 @@
     <_TargetFrameworkDirectories>$(MSBuildThisFileDirectory)/Documentation</_TargetFrameworkDirectories>
     <_FullFrameworkReferenceAssemblyPaths>$(MSBuildThisFileDirectory)/Documentation</_FullFrameworkReferenceAssemblyPaths>
     <SkipCommonResourcesIncludes>true</SkipCommonResourcesIncludes>
-    <LangVersion>7.2</LangVersion>
+    <LangVersion>7.3</LangVersion>
```
Comment (jkotas) [inline src/mscorlib/System.Private.CoreLib.csproj]: We may just want to do `latest` here.

**Comment by kasper3** [comment]
https://ci.dot.net/job/dotnet_coreclr/job/master/job/x64_checked_osx10.12_innerloop_prtest/2445/consoleText
```
FAILED: file_io/GetFileSize/test1/paltest_getfilesize_test1. Exit code: 1
.GetFileSizeEx: ERROR -> SetEndOfFile failed due to lack of disk space
```

File: src/mscorlib/System.Private.CoreLib.csproj (hunk 1 above)
Comment (weshaggard) [reply src/mscorlib/System.Private.CoreLib.csproj]: We shouldn't do this until we are done with 2.1. We don't want to depend on 7.3 yet as it is still under development.

Comment (jkotas) [reply src/mscorlib/System.Private.CoreLib.csproj]: I do not think we will take this PR for 2.1 for other reasons.

But I do not see a fundamental problem with always using latest.

Comment (weshaggard) [reply src/mscorlib/System.Private.CoreLib.csproj]: >I do not think we will take this PR for 2.1 for other reasons.

If that is the case we shouldn't merge this until after our final merge into release/2.1 which is scheduled for this friday.

**Comment by jkotas** [comment]
@dotnet-bot test this please

**Pull request merged**